### PR TITLE
feat: add ability for drivers to update server and add method maps like plugins

### PR DESCRIPTION
**Pull request by jlipps**
there's no reason that only plugins should be able to do these things. drivers would reasonably want to add new commands too.

* depends on https://github.com/appium/appium-base-driver/pull/459 (for the new interface)
* depends on https://github.com/appium/appium-fake-driver/pull/71 (for testing)
* depends on https://github.com/appium/appium-plugins/pull/6 (for testing)

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
...
 }
 
+function getActivePlugins (args, pluginConfig) {
+  // find any plugin name which has been installed, and which has been requested for activation by
+  // using the --plugins flag, and turn each one into an instantiated plugin object, so we can send
+  // them as objects to the server init. we also want to send/assign them to the umbrella driver so
+  // it can use them to wrap command execution
+  return Object.keys(pluginConfig.installedExtensions).filter((pluginName) =>
+    _.includes(args.plugins, pluginName) ||
+    (args.plugins.length === 1 && args.plugins[0] === USE_ALL_PLUGINS)
```
Comment (mykola-mokhnach) [inline lib/main.js:99]: is args.plugins always defined?

File: lib/main.js [hunk 2]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
...
+  ).map((pluginName) => {
+    try {
+      logger.info(`Attempting to load plugin ${pluginName}...`);
+      const PluginClass = pluginConfig.require(pluginName);
+      PluginClass.pluginName = pluginName; // store the plugin name on the class so it can be used later
+      return PluginClass;
+    } catch (err) {
+      logger.error(`Could not load plugin '${pluginName}', so it will not be available. Error ` +
```
Comment (mykola-mokhnach) [inline lib/main.js:107]: err.message

File: lib/main.js [hunk 3]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
...
+  return Object.keys(pluginConfig.installedExtensions).filter((pluginName) =>
+    _.includes(args.plugins, pluginName) ||
+    (args.plugins.length === 1 && args.plugins[0] === USE_ALL_PLUGINS)
+  ).map((pluginName) => {
+    try {
+      logger.info(`Attempting to load plugin ${pluginName}...`);
+      const PluginClass = pluginConfig.require(pluginName);
+      PluginClass.pluginName = pluginName; // store the plugin name on the class so it can be used later
+      return PluginClass;
+    } catch (err) {
```
Comment (mykola-mokhnach) [inline lib/main.js:106]: it makes sense to also print the error stack to the debug log

File: lib/main.js [hunk 4]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
   logger.info(logMessage);
 }
 
+function getActivePlugins (args, pluginConfig) {
```
Comment (mykola-mokhnach) [inline lib/main.js:92]: is this function called once?

File: lib/main.js [hunk 5]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
...
+  }).filter(Boolean);
+}
+
+function getActiveDrivers (driverConfig) {
+  return Object.keys(driverConfig.installedExtensions).map((driverName) => {
+    try {
+      logger.info(`Attempting to load driver ${driverName}...`);
+      return driverConfig.require(driverName);
+    } catch (err) {
+      logger.error(`Could not load driver '${driverName}', so it will not be available. Error ` +
```
Comment (mykola-mokhnach) [inline lib/main.js:120]: same as above

File: lib/main.js [hunk 6]
```diff
@@ -89,6 +89,52 @@ function logServerPort (address, port) {
   logger.info(logMessage);
 }
 
+function getActivePlugins (args, pluginConfig) {
+  // find any plugin name which has been installed, and which has been requested for activation by
```
Comment (mykola-mokhnach) [inline lib/main.js:93]: why not to turn that into a proper function docstring?

File: lib/main.js [hunk 7]
```diff
@@ -133,53 +179,50 @@ async function main (args = null) {
...
   };
   if (args.keepAliveTimeout) {
     serverOpts.keepAliveTimeout = args.keepAliveTimeout * 1000;
   }
-  let server = await baseServer(serverOpts);
+  let server;
+  try {
+    server = await baseServer(serverOpts);
+  } catch (err) {
+    logger.error(`Could not configure Appium server. It's possible that a driver or plugin tried ` +
```
Comment (mykola-mokhnach) [inline lib/main.js:213]: please add the error stack to the debug log

File: lib/main.js [hunk 8]
```diff
@@ -133,53 +179,50 @@ async function main (args = null) {
...
+                 `to update the server and failed. Original error: ${err}`);
```
Comment (mykola-mokhnach) [inline lib/main.js:214]: err.message

File: lib/main.js [hunk 9]
```diff
@@ -133,53 +179,50 @@ async function main (args = null) {
...
+    process.exit(1);
+    return;
```
Comment (mykola-mokhnach) [inline lib/main.js:216]: return process.exit(1);

File: lib/appium.js [hunk 10]
```diff
@@ -355,10 +359,37 @@ class AppiumDriver extends BaseDriver {
     }
   }
 
-  pluginsToHandleCmd (cmd) {
+  pluginsToHandleCmd (cmd, sessionId) {
+    // to get plugins for a command, we either get the plugin instances associated with the
```
Comment (mykola-mokhnach) [inline lib/appium.js:363]: how about turning this into a docstring?

File: test/driver-e2e-specs.js [hunk 11]
```diff
@@ -266,6 +274,17 @@ describe('FakeDriver - via HTTP', function () {
 
       createSessionStub.restore();
     });
+
+    it('should allow drivers to update the method map with new routes and commands', async function () {
+      let driver = await wdio({...wdOpts, capabilities: caps});
+      const {sessionId} = driver;
+      try {
+        await axios.post(`${baseUrl}/${sessionId}/fakedriver`, {thing: {yes: 'lolno'}});
```
Comment (KazuCocoa) [inline test/driver-e2e-specs.js:282]: 👍

File: lib/main.js (hunk 1 above)
Comment (jlipps) [reply lib/main.js:99]: yes, defaults to `[]`

**Comment by jlipps** [comment]
all right, CI passed with all the dependencies published

**Pull request merged**